core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1207 @@
+
+    expect(sgls.getLastComputedArrangement()).toEqual(computedLayout);
+  });
+
+  it('should return graph boundaries with width less than equal to' +
+    ' maximum allowed graph width', () => {
+    let nodeData = {
+      State5: {
+        depth: 2,
+        offset: 1,
+        reachable: true,
+        x0: 0.29750000000000004,
+        y0: 0.54,
+        xLabel: 0.3875,
+        yLabel: 0.6,
+        id: 'State5',
+        label: 'State5',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State4: {
+        depth: 1,
+        offset: 3,
+        reachable: true,
+        x0: 0.7475,
+        y0: 0.33999999999999997,
+        xLabel: 0.8375,
+        yLabel: 0.4,
+        id: 'State4',
+        label: 'State4',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State3: {
+        depth: 1,
+        offset: 2,
+        reachable: true,
+        x0: 0.5225000000000001,
+        y0: 0.33999999999999997,
+        xLabel: 0.6125,
+        yLabel: 0.4,
+        id: 'State3',
+        label: 'State3',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State2: {
+        depth: 1,
+        offset: 1,
+        reachable: true,
+        x0: 0.29750000000000004,
+        y0: 0.33999999999999997,
+        xLabel: 0.3875,
+        yLabel: 0.4,
+        id: 'State2',
+        label: 'State2',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State1: {
+        depth: 1,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.33999999999999997,
+        xLabel: 0.1625,
+        yLabel: 0.4,
+        id: 'State1',
+        label: 'State1',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      Introduction: {
+        depth: 0,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.14,
+        xLabel: 0.1625,
+        yLabel: 0.2,
+        id: 'Introduction',
+        label: 'Introduction',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      End: {
+        depth: 3,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.7400000000000001,
+        xLabel: 0.1625,
+        yLabel: 0.8,
+        id: 'End',
+        label: 'End',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      }
+    };
+    let graphWidth = sgls.getGraphWidth(
+      AppConstants.MAX_NODES_PER_ROW, AppConstants.MAX_NODE_LABEL_LENGTH);
+    let graphHeight = sgls.getGraphHeight(nodeData);
+    let nodeDataWithPositionValueInPixel = sgls.modifyPositionValues(
+      nodeData, graphWidth, graphHeight);
+    let expectedGraphBoundaries = {
```

Comment:
Can you add some explanation (in comments) for why these are the correct values? In general, use comments to explain the "why" behind anything that is not self-evident.

Reply:
Added comment

Reply:
I don't see a comment for where the values in expectedGraphBoundaries comes from. I see an explanation for nodeDataWithPositionValueInPixel but that is a different thing.

There should be an explanation for why the values in expectedGraphBoundaries are correct, which can be verified directly based on an examination of the nodeData in this test.

Reply:
Added a comment, though the values in `expectedGraphBoundaries` are calculated from `nodeDataWithPositionValueInPixel`, so they not be directly verifiable from `nodeData`. Should I add `nodeDataWithPositionValueInPixel` manually?

Reply:
Ah -- I guess I'm confused about what this test is testing. If the main thing that's being tested is getGraphBoundaries() then yes, just add a custom nodeDataWithPositionValueInPixel here and drop all reference to nodeData. Try to make your custom nodeDataWithPositionValueInPixel easy to understand so that the expected output can be verified to follow from it (if this is not clear, add comments to explain).

Reply:
Done